lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    StringRef outputFile, const SupplementaryOutputPaths &pathsFromArguments,
+    const InputFile &input) const {
+  StringRef implicitBasis = deriveImplicitBasis(outputFile, input);
+
+  using namespace options;
+
+  auto dependenciesFilePath = determineSupplementaryOutputFilename(
+      OPT_emit_dependencies, pathsFromArguments.DependenciesFilePath, "d", "",
+      implicitBasis);
+  if (!dependenciesFilePath)
+    return None;
+
+  auto referenceDependenciesFilePath = determineSupplementaryOutputFilename(
+      OPT_emit_reference_dependencies,
+      pathsFromArguments.ReferenceDependenciesFilePath, "swiftdeps", "",
+      implicitBasis);
+  if (!referenceDependenciesFilePath)
+    return None;
+
+  auto serializedDiagnosticsPath = determineSupplementaryOutputFilename(
+      OPT_serialize_diagnostics, pathsFromArguments.SerializedDiagnosticsPath,
+      "dia", "", implicitBasis);
+
+  if (!serializedDiagnosticsPath)
+    return None;
+
+  auto objCHeaderOutputPath = determineSupplementaryOutputFilename(
+      OPT_emit_objc_header, pathsFromArguments.ObjCHeaderOutputPath, "h", "",
+      implicitBasis);
+  if (!objCHeaderOutputPath)
+    return None;
+
+  auto loadedModuleTracePath = determineSupplementaryOutputFilename(
+      OPT_emit_loaded_module_trace, pathsFromArguments.LoadedModuleTracePath,
+      "trace.json", "", implicitBasis);
+  if (!loadedModuleTracePath)
+    return None;
+
+  auto tbdPath = determineSupplementaryOutputFilename(
+      OPT_emit_tbd, pathsFromArguments.TBDPath, "tbd", "", implicitBasis);
+  if (!tbdPath)
+    return None;
+
+  auto moduleDocOutputPath = determineSupplementaryOutputFilename(
+      OPT_emit_module_doc, pathsFromArguments.ModuleDocOutputPath,
+      SERIALIZED_MODULE_DOC_EXTENSION, "", implicitBasis);
+  if (!moduleDocOutputPath)
+    return None;
+
+  ID emitModuleOption;
+  std::string moduleExtension;
+  std::string mainOutputIfUsableForModule;
+  deriveModulePathParameters(emitModuleOption, moduleExtension,
+                             mainOutputIfUsableForModule);
+
+  auto moduleOutputPath = determineSupplementaryOutputFilename(
+      emitModuleOption, pathsFromArguments.ModuleOutputPath, moduleExtension,
+      mainOutputIfUsableForModule, implicitBasis);
+  if (!moduleOutputPath)
+    return None;
+
+  return SupplementaryOutputPaths(
+      *objCHeaderOutputPath, *moduleOutputPath, *moduleDocOutputPath,
+      *dependenciesFilePath, *referenceDependenciesFilePath,
+      *serializedDiagnosticsPath, *loadedModuleTracePath, *tbdPath);
+}
+
+StringRef SupplementaryOutputPathsComputer::deriveImplicitBasis(
+    StringRef outputFilename, const InputFile &input) const {
+  if (!outputFilename.empty() && outputFilename != "-")
+    // Put the serialized diagnostics file next to the output file.
+    return outputFilename;
+
+  // If we have a primary input, so use that as the basis for the name of the
+  // serialized diagnostics file, otherwise fall back on the
```

Comment:
Comment needs updating.

Reply:
Good catch, thanks! Much of the comments in that function were redundant, so I streamlined them.